### PR TITLE
Improve throughput on arm64

### PR DESCRIPTION
**Pull request by ncruces**
Hi! 👋

This branch includes a slew of updates that significantly improve performance on [`arm64`](https://github.com/ncruces/adiantum/actions/runs/9009610320/job/24754159188).

Before:
```
BenchmarkAdiantum/XChaCha12_Encrypt-3        	  113114	     11517 ns/op	 355.66 MB/s	       0 B/op	       0 allocs/op
BenchmarkAdiantum/XChaCha12_Decrypt-3        	  105201	     11246 ns/op	 364.22 MB/s	       0 B/op	       0 allocs/op
```

After:
```
BenchmarkAdiantum/XChaCha12_Encrypt-3        	  285481	      4209 ns/op	 973.20 MB/s	       0 B/op	       0 allocs/op
BenchmarkAdiantum/XChaCha12_Decrypt-3        	  285500	      4224 ns/op	 969.65 MB/s	       0 B/op	       0 allocs/op
```

You may want to consider each commit individually, or reject this altogether. I really don't mind doing some additional work to upstream this! This is just where I'm at, and I understand some of it is opinionated.

Commits:
1. 5028739f27d5118ca4212b2f13b86f15840f473f updates dependencies, and set Go to 1.20 (needed by c8607862f781328052ea21cb93d4656cf6ae6dce)
2. 852cde011c6466d058acf2f5a2ad27a000301876 `github.com/aead/chacha20` is [archived](https://github.com/aead/chacha20/), vendor it
3. 878546e283872a5f29eea11cc7690f6ea2e521d0 removes the bits that aren't used from 852cde011c6466d058acf2f5a2ad27a000301876
4. 8ae39ea1182fe34430331278836d46eb8549464c updates to use `go:build` rather than `+build`, and revises tags (`nacl` doesn't exist anymore, etc)
5. 640c72074dea998ad2a166a365d73bb7e721a7a3 adapts `arm64` assembly from `golang.org/x/crypto/chacha20`
6. e3261e2e06f26ae14c0fe981d692533cd6cc8900 refactor `chacha` for code reuse
7. b6059c569cae0e49e810023b54cd803d77367d40 improves the fallback version of `nh.Sum`
8. c8607862f781328052ea21cb93d4656cf6ae6dce adds little endian version of `nh.Sum`
9. cb1b8cc2c8e186be76d79373f1a262a54dbe2dd5 avoids deprecated `ioutil`

Commit 640c72074dea998ad2a166a365d73bb7e721a7a3 is a bit hacked up, and might benefit from additional work. I just copied the assembly, and made the rounds configurable. Didn't even bother with fixing the setup parts. OTOH, this means that if upstream has fixes, they're easier to integrate. Also, the same interface is used by `ppc64` and `s390x` assembly.

Commit c8607862f781328052ea21cb93d4656cf6ae6dce is a bit of a cop out. Instead of sitting down and writing `arm64` assembly, I just coded a Go version that should work on any little endian architecture, and used `unsafe` for setup. Still, it's about as fast `sumAsm` on `amd64`, and produces decent code on a wide range of architectures. This is responsible for much of the speed up on Adiantum.

Commit b6059c569cae0e49e810023b54cd803d77367d40 improves the situation a bit for the big endian architectures left out in c8607862f781328052ea21cb93d4656cf6ae6dce, by unrolling a loop and avoiding some bounds checks.

### CODE REVIEW

**Comment by lukechampine** [comment]
whoa! Amazing work, will dig into this soon. Thank you!

**Comment by ncruces** [comment]
Feel free to ask me to split this up in PRs that make sense to you, and that you can review independently.

File: internal/chacha/chacha_arm64.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	binary.LittleEndian.PutUint32(state[4:], sigma[1])
+	binary.LittleEndian.PutUint32(state[8:], sigma[2])
+	binary.LittleEndian.PutUint32(state[12:], sigma[3])
+	copy(state[16:], key[:])
+	copy(state[48:], nonce[:])
+}
+
+func xorKeyStream(dst, src []byte, block, state *[64]byte, rounds int) int {
+	switch {
+	case useVX:
```
Comment (lukechampine) [inline internal/chacha/chacha_arm64.go:15]: do we even need a switch case if `useVX` will always be true?

File: internal/chacha/chacha.go [hunk 2]
```diff
@@ -57,13 +57,6 @@ func setup(state *[64]byte, nonce, key []byte) (err error) {
 		HChaCha20(&tmpKey, &hNonce, &tmpKey)
 		copy(Nonce[8:], nonce[16:])
 		initialize(state, tmpKey[:], &Nonce)
-
-		// BUG(aead): A "good" compiler will remove this (optimizations)
-		//			  But using the provided key instead of tmpKey,
-		//			  will change the key (-> probably confuses users)
-		for i := range tmpKey {
-			tmpKey[i] = 0
-		}
```
Comment (lukechampine) [inline internal/chacha/chacha.go]: any reason to remove this? It probably isn't being optimized away, and zeroing keys is good practice.

**Review by lukechampine — COMMENTED** [review]
I went commit-by-commit and it was pretty easy to digest that way. Looks to be ~2.5-3x speedup on my M2 as well -- great stuff!

File: internal/chacha/chacha_arm64.go (hunk 1 above)
Comment (ncruces) [reply internal/chacha/chacha_arm64.go:15]: This is implemented in the same vein as the similar flags on x86 version the package.

The variable can be changed in tests to ensure the version without assembly can be tested too.

I actually considered doing the same for the assembly on NH package.

File: internal/chacha/chacha.go (hunk 2 above)
Comment (ncruces) [reply internal/chacha/chacha.go]: Well, yes, it wasn't optimized away.

And the way it's used within the package, it's a temporary key, derived from another temporary key, all stored in the stack.

This was the only key that's being zeroed, so it seemed redundant (as we're not taking care to do the same elsewhere). But I'll revert it.

Depending on your willingness to upgrade to 1.21, we can use `clear`.

File: internal/chacha/chacha_arm64.go (hunk 1 above)
Comment (lukechampine) [reply internal/chacha/chacha_arm64.go:15]: ah, makes sense

File: internal/chacha/chacha.go (hunk 2 above)
Comment (lukechampine) [reply internal/chacha/chacha.go]: I'm ambivalent; mostly I don't want to meddle with the original package too much, but maybe that ship has sailed. You're right that it's not very consistent to care about zeroing here and not in the rest of the package. (And frankly I've never been too concerned about key material floating around in program memory.)

**Comment by lukechampine** [comment]
Thanks again for this. I hope Go (and [avo](https://github.com/mmcloughlin/avo)) get better support for ARM assembly soon; I would very much like to accelerate `blake3` in the same way.

**Comment by ncruces** [comment]
Awesome! It'd be great if you could tag a 1.1 or 1.0.1 (I don't know which one applies).

**Comment by lukechampine** [comment]
v1.0.1 seems disrespectful for such a big improvement, so I'll do v1.1.0 :P

**Pull request merged**